[이찬주] Week4

ckswnskfk opened the pull request:
## 요구사항

### 공통
- [x] 이메일 input에서 focus out 할 때, 값이 없을 경우 input에 빨강색 테두리와 아래에 “이메일을 입력해주세요.” 빨강색 에러 메세지가 보입니다
- [x] 이메일 input에서 focus out 할 때, 이메일 형식에 맞지 않는 값이 있는 경우 input에 빨강색 테두리와 아래에 “올바른 이메일 주소가 아닙니다.” 빨강색 에러 메세지가 보입니다
- [x] 비밀번호 input에서 focus out 할 때, 값이 8자 미만으로 있거나 문자열만 있거나 숫자만 있는 경우, input에 빨강색 테두리와 아래에 “비밀번호는 영문, 숫자 조합 8자 이상 입력해 주세요.” 빨강색 에러 메세지가 보입니다.
- [x] 비밀번호 input에서 focus out 할 때, 값이 없을 경우 아래에 “비밀번호를 입력해주세요.” 에러 메세지가 보입니다.
- [x] 이메일, 비밀번호 input에 에러 관련 디자인을 Components 영역의 에러 케이스로 적용합니다.

### 회원가입 페이지
- [x] 이메일 input에서 focus out 일 때, input 값이 test@codeit.com 일 경우 input에 빨강색 테두리와 아래에 “이미 사용 중인 이메일입니다.” 빨강색 에러 메세지가 보입니다.
- [x] 비밀번호 input과 비밀번호 확인 input의 값이 다른 경우, 비밀번호 확인 input에 빨강색 테두리와 아래에 “비밀번호가 일치하지 않아요.” 빨강색 에러 메세지가 보입니다.
- [x] 회원가입을 실행할 경우, 문제가 있는 경우 문제가 있는 input에 빨강색 테두리와 에러 메세지가 보입니다.
- [x] 이외의 유효한 회원가입 시도의 경우, “/folder”로 이동합니다.
- [x] 회원가입 버튼 클릭 또는 Enter키 입력으로 회원가입 됩니다.

### 로그인 페이지
- [x] 이메일: test@codeit.com, 비밀번호: codeit101 으로 로그인 시도경우, “/folder” 페이지로 이동합니다.
- [x] 이외의 로그인 시도의 경우 이메일, 비밀번호 input에 빨강색 테두리와 각각의 input아래에 “이메일을 확인해주세요.”, “비밀번호를 확인해주세요.” 빨강색 에러 메세지가 보입니다.

### 심화
- [x] 눈 모양 아이콘 클릭시 비밀번호의 문자열이 보이기도 하고, 가려지기도 합니다.
- [x] 비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보이도록 합니다.
- [x] 로그인, 회원가입 페이지에 공통적으로 사용하는 로직이 있다면, 반복하지 않고 공통된 로직을 모듈로 분리해 사용합니다.

## 주요 변경사항

- 회원가입, 로그인 페이지 별 js 파일(signup.js, signin.js) 작성
- 공통 기능 js 파일(sign.js) 작성
- dom 조작 공통 함수 파일(dom.js) 작성
- 어플리케이션 내부에서 전체적으로 쓸 수있는 함수를 담아놓은 파일(util.js) 작성
- user 데이터 보관하는 파일 작성(users.js)

## 스크린샷

<img width="461" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/df97c6b0-99ba-4f76-b9f2-55dc1c021c71">
<img width="486" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/484ee338-1877-42c1-af81-660e82aeea96">
<img width="491" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/cfc380ea-a0de-4dff-a715-d13cc3dc1808">
<img width="468" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/e0721437-4baa-4b06-9f93-984013caccb4">
<img width="460" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/2b87f0fc-9cb3-407d-8316-0d3877eba731">
<img width="465" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/ab2af0fe-fd82-489a-b1e2-06c00d3940e9">
<img width="439" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/a19e5d9f-a07f-42a8-8bc3-fbac89eac53a">
<img width="452" alt="image" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/98e55971-613f-483e-870c-f61e88495a7e">
<img width="467" alt="로그인 시도 시 회원이 아닌 경우" src="https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/44164404/a684efee-a918-4744-a897-d082523a5154">


## 멘토에게

- 폴더 구조를 어떻게 쉽고 효율적으로 잡을 수 있을까요?
- 모듈화를 한다고 해봤으나(그 과정에서 너무 뇌절한 게 아닌가 싶습니다..:cry:) 여전히 막연하지만 확장성은 낮고 결합도가 높아보입니다. 어떻게 개선할 수 있을까요?
- 함수들이 자신이 포함된 파일에 적절하게 들어가 있는 지 궁금합니다.
- 함수 내부 타입 검사가 필요해 보입니다. 타입스크립트를 쓰면 해결되는 문제인가요?
- 함수명을 간결하고 한 눈에 의미를 할 수 있게 짓고 싶습니다. 함수명에 대한 조언도 부탁드립니다!

ckswnskfk committed:
Style: 중복된 title 태그 삭제

ckswnskfk committed:
Fix: 폰트 css preload 경고 해결, sign 페이지 경고 메시지 마크업 변경

preload 했지만 사용되지 않는다는 경고 발생. 하지만 실제론 html 전체에 해당 폰트 적용 중 -> preload 속성 뺌

ckswnskfk committed:
Design: 접근성을 지키며 요소 숨기는 스타일 추가, sign 페이지 경고 스타일 추가

ckswnskfk committed:
Design: 각각의 페이지에 해당하는 js파일 생성 등

숨김 스타일 클래스명 변경
signup 페이지 기능 구현을 위한 input 타입 변경

ckswnskfk committed:
Feat: 회원가입 페이지 인풋 박스 기능 구현 완성

기능 별 파일 생성 및 코드 분리

ckswnskfk committed:
Feat: 로그인 페이지 인풋박스 기능 구현 완성

ckswnskfk committed:
Feat: 회원가입 버튼 기능 완성

모든 validating 통과 시 folder.html로 이동
validating 함수 리턴 값 설정(boolean)
일관성을 위한 함수명 수정

ckswnskfk committed:
Feat: 비밀번호 input 눈 모양 기능 완성

모듈화를 위한 함수 위치 이동
몇몇 함수 더 알맞는 파일로 이동

ckswnskfk committed:
Feat: 멤버가 아닌 경우 인풋 박스에 경고 메시지 출력 기능 완성

ckswnskfk committed:
Refactor: 중복된 기능의 함수 추상화

ckswnskfk committed:
Refactor: export 구문 파일의 최하단 작성

함수 옆에 export문이 있으면 외부에서 사용되는 지 알기 어려움

sstaar91 commented on scripts/dom.js:
```diff
@@ -0,0 +1,29 @@
+// 공통
```
DOM에 접근하는 코드를 이렇게 분리하셨네요!
이렇게 분리하게 되면 DOM에 접근하게 되는 상황에서 골라서 사용할 수 있다는 장점이 있지만,
"개인적인 생각으로는"  너무 잘게 쪼갠 모듈화는 오히려 그 코드를 직접 사용하는 것보다 
복잡한 과정을 거치게 될 수 있다라는 생각입니다.

현재 이 파일에 작성된 코드를 봤을 때, 함수를 생성하는 것 보다
안에 있는 코드를 직접 사용하는게 더 직관적이지 않을까 하는 생각이 들었습니다.

물론 함수의 이름을 통해서 어떤 역할을 하는 함수인지 파악할 수 있지만,
해당 함수의 이름만 가지고 짐작하는 것 보다 해당 함수의 로직을 파악하는 것이 더 정확하기에
결국 해당 함수를 사용하려면 이 파일을 확인해야 하는 과정이 발생할 수 있게 될 것 같습니다!

sstaar91 commented on scripts/sign.js:
```diff
@@ -0,0 +1,167 @@
+    ".input-error__message"
+  );
+  addElementClass(errorMessageSpan, "hide");
+}
```
이 두개 함수를 합쳐서 handleErrorMessage로 만들면 어떨까요?
어떠한 요소를 특정한 뒤, 파라미터로 해당 함수가 실행되는 상황을 전달하면
하나의 함수로 비슷한 로직의 동작을 구성할 수 있을 것 같습니다.

sstaar91 commented on scripts/sign.js:
```diff
@@ -0,0 +1,167 @@
+    return "";
+  }
+  return message;
+}
```
이 3개의 함수는 이메일에 입력된 값을 체크하는 함수로 보여집니다.

입력값이 비었을때 / 적합한 이메일의 형식이 아닐때 / 입력한 이메일이 맞지 않는 이메일 일때
위 3가지 상황을 각각 함수로 나눠서 관리하고 계시는데

제 생각으로는 이 3개의 함수 모두 동일한 로직으로 구성이 되어 있습니다
target의 value가 어떠한 조건에 참일경우는 "", 거짓일 경우는 message를 리턴하는 로직이 동일합니다.

이런 경우에는 early return을 이용해서 하나의 함수에서 처리하는 것이 더 효율적이라고 생각합니다.

sstaar91 commented on scripts/signin.js:
```diff
@@ -0,0 +1,120 @@
+      validator: checkEmailValid,
+      message: "올바른 이메일 주소가 아닙니다.",
+    },
+  ];
```
어떠한 데이터를 다룰 때 배열과 객체를 이용하는 방법은 제가 좋아하는 방법입니다👍
이렇게 데이터를 관리하게 되면 하드코딩 하게 되는 상황은 줄어들고,
데이터의 추가가 있다고 하더라도 해당 객체 혹은 배열만 변경하면 되는거라
유지보수에도 좋다라고 생각합니다 ㅎㅎ

sstaar91 commented on scripts/signin.js:
```diff
@@ -0,0 +1,120 @@
+  ".form__password .form__input--eye-off"
+);
+
+passwordInput.addEventListener("focusout", onPasswordFocusoutValid);
```
<함께 생각할 부분>

개인적으로는 자바스크립트 내에서도 코드의 컨벤션이 있다고 생각합니다.
예를 들어 현재 찬주님께서는 역할에 따라서 DOM, 이벤트리스너, 사용하는 함수 의 순서로 코드를 배치하셨는데
저는 주로 DOM에 접근하는 부분끼리, 이벤트 리스너끼리, 함수끼리 묶어서 배치하는 형태로 구현하곤 합니다.

어떤 컨벤션이 옳다라는 건 아니지만, 나중에 팀 프로젝트를 진행할 때는 
이러한 부분도 팀원과 상의하는 것이 좋을 것 같아 살짝 남겨봅니다!

ckswnskfk replied:
생각할 거리 주셔서 감사합니다!
저는 관련 있는 코드는 가까이 있는 게 이해하기 쉽다고 들어서 사용되는 곳 근처에서 쓰려고 노력했습니다ㅎㅎ

sstaar91 reviewed (commented):
찬주님 수고하셨습니다!👍
처음으로 자바스크립트를 이용해 몇가지 기능들을 구현해 보셨는데,
이론과 실제를 적용하는 부분이다 보니까 다소 어렵고 힘들게 느껴질 수 있을거에요!
그럼에도 불구하고 찬주님께서 작성하신 코드는 굉장한 고민과 노력들이 담겨 있다고 생각합니다!!

질문 주신 내용에 대해서 답변을 드리면

[폴더 구조에 대해]

폴더 구조는 일단 크게로는 각 파일의 타입에 따라 나눌 수 있을 것 같습니다.
html 파일은 html 파일끼리, css는 css 끼리 이런 식으로요

그 다음으로는 해당 파일의 역할에 따라 나눌 수 있을 것 같습니다.
css 파일 중에서도 전체 페이지에 공통으로 적용되는 스타일 인지 아닌지
js 파일 중에서도 여러 방면에 사용할 수 있는 모듈인지 아닌지

그런데 예외적으로는 이런 경우도 있습니다.
만약 signup과 관련된 파일 (signup.html, css, js) 가 있다고 가정했을 때,
이 3개의 파일은 각각 분리해서 관리하는 것이 아닌 하나의 파일 안에서 같이 관리하는게
차후 유지보수를 위해서 좋은 방법입니다.

[함수가 적절하게 분리 되어있는지?]

이 부분은 dom, sign, utils 파일에 대한 질문이라고 생각하고 답변을 드리면
적절하게 분리가 되어있다고 볼 수 있습니다.

[함수 내부 타입 검사]

함수 내부에서 typeof를 이용해 선제적으로 처리를 하는 방법도 있을 수 있겠습니다.
물론 타입스크립트를 이용하는 방법이 더 효과적이겠지만,
지금 상황에서는 내부 타입에 대한 검사가 필요할 때는 typeof를 사용해서
전달받은 데이터의 타입을 먼저 체크한 뒤 로직이 이어지도록 구성하는 것이 좋겠네요

[함수명 짓기]

이 부분은 저도 항상 고민하고, 어려워 하는 부분이기 때문에 어떻게 해야한다! 고 
명확하게 말씀드릴 순 없지만 참고가 될 만한 글의 링크는 전달드릴게요! ([변수명 짓기](https://brunch.co.kr/@wapj2000/29)) 
너무 함수명이 길다면 범용적으로 사용되는 단어의 약어를 활용하는 것도 하나의 방법이 되겠습니다.

[모듈화]

찬주님 코드에서 가장 얘기를 나누고 싶었던 부분입니다.
모듈화라는 건 결국 코드의 응집도를 높이고 결합도를 낮추기 위한 하나의 방법입니다.
그래서 찬주님께서 작성하신 코드는 이러한 부분을 많이 고려해서 작성하신 것으로 보이는데
너무 높은 추상화는 오히려 코드의 가독성과 유지보수를 낮추는 경향이 있을 수 있습니다.

함수는 하나의 역할을 해야한다 라는 말은 언제나 저도 코드를 작성할 때 생각하는 부분이지만
하나의 역할이라고 하는것이 조금은 모호하게 느껴질 수 있을 것 같습니다.
제가 생각하는 하나의 역할은 하나의 기능 단위라고 생각합니다.

예를 들어서 이번 week4의 과제를 기준으로 한다면
input 에 입력된 값이 어떠한 조건에 부합하지 않는 경우
에러메세지를 보여주고 input의 스타일을 변경한다 라는 것이 기능에 대한 설명이라면
조건을 체크하는 함수 / 요소를 추가하거나 변경하는 함수 로 나눌 수 있을 것 같습니다.

현재 찬주님의 코드에서는 세세하게 하나하나의 동작들을 나눠놔서
추상화는 굉장히 높지만 함수가 잘게 쪼개져 있어서 오히려 코드의 가독성에 영향을 줄 수 있을 것 같네요!

ckswnskfk committed:
Rename: 폴더 구조 변경

랜딩 페이지 경로 더 깔끔하게 수정

ckswnskfk committed:
Refactor: 한 줄 짜리 dom 조작 함수 삭제

한 줄 짜리 코드임에도 나누어져 있던 함수를 삭제하고 해당 내용을 함수 내에 직접 작성

ckswnskfk committed:
Refactor: 비슷한 동작을 하는 함수 합침

showErrorMessage 함수와 hideErrorMessage 함수를 합쳐 handleErrorMessage 함수 생성
input값 체크하는 로직 합침

ckswnskfk committed:
Refactor: 가독성을 해치는 함수 validatingMachine 삭제

validatingMachine 함수의 기능은 해당 함수를 호출하는 부분에 직접 작성

ckswnskfk committed:
Refactor: 비슷한 동작의 함수 하나로 합침

emailValidationFailed, passwordValidationFailed 함수 합침
emailValidationSucceeded, passwordValidationSucceeded 함수 합침
checkIsOurMember 함수 대신 해당 함수 내부에서 호출하는 함수 직접 사용

ckswnskfk commented:
@sstaar91 멘토님 리뷰해 주신 내용 최대한 반영하려고 노력했습니다!
폴더 구조도 말씀해주신 내용 바탕으로 반영해봤는데, url이 깔끔하지 않은 것 같기도 하고 아직 괜찮은지 모르겠습니다ㅎㅎ
한 번 봐주셨으면 좋겠어요 :bow:
멘토링 때 보여주신 것처럼 가독성 좋게 한 번에 되지는 않네요!
작성한 제가 봐도 알아보기 힘든 코드 리뷰해주셔서 감사해요 😄

sstaar91 commented on pages/member/sign.js:
```diff
@@ -0,0 +1,115 @@
+  },
+};
+
+function handleErrorMessage(element, errorMessage = "", addOrRemoveHide) {
```
```suggestion
function handleErrorMessage(element, addOrRemoveHide, errorMessage = "") {
```

파라미터에 기본값을 설정하면, 파라미터의 가장 뒷 부분에 위치하는 것이 일반적입니다.
이렇게 될 경우 실제 호출시 아규먼트의 생략도 가능하기 때문에
조금 더 가독성 높은 코드를 작성할 수 있습니다.

sstaar91 reviewed (commented):
찬주님 수고하셨습니다!

딱 폴더구조가 제가 주로 사용하는 폴더 구조의 형태라서 마치 제가 구분을 해 놓은 것 같네요 ㅋㅋㅋ
물론 저 구조가 답이다라고는 할 수 없기에 더 나은 구조에 대해서는
직접 찾아보고 구성해 보는 것도 좋겠습니다 ㅎㅎ

그리고 멘토링 시간을 거쳐서 제가 말씀드렸던 부분들이 훌륭하게 반영되어 있어서 감동했습니다.!
잘 반영해 주셔서 감사해요!

한 가지 리뷰 사항도 남겨드렸으니 이 부분도 확인해주세요~!

ckswnskfk committed:
Refactor: 기본값 설정한 파라미터 가장 뒤로 이동

ckswnskfk committed:
Fix: 리팩토링에 따른 파라미터 순서 변경으로 동작하지 않던 함수 순서 맞춤
